### PR TITLE
added hook for dashboard filter front

### PR DESCRIPTION
**Pull request by bessantoy**
Hello, 

I am a developer at IT Gouvernance and we develop several plugins on GLPI. There is a problem that we regularly encountered, in fact, when we wanted to add a filter using the DASHBOARD_FILTERS hook, we also had to modify the Filter.php file in order to write a static function there that calls Filter::displayList so that the dropdown can be displayed on the dashboard. To overcome this problem I added a new Hook and modified the Grid:: getFilterHtml function so that we can provide the front of the filter directly from the plugin. We tested this feature and it works.

Regards,

BESSIERES Anthony from IT Gouvernance

### CODE REVIEW

**Comment by bessantoy** [comment]
Hello, 
I allow myself to remind you about this pull request, please consider it.

**Review by orthagh — COMMENTED** [review]
I agree with hook addition proposal.
But i found the parameter and the return odd, the core function ask the html of a specific filter_id, and the hook return all the filter and it's required to check the specific filter exist in the hook.

If i would have develop myself the hook, i would have set the same input/output as the above core call

So 2 parameters:
- string $filter id // name of the filter
- mixed (string/array) $filter_values // value(s) send when a filter is updated in the grid

And the return should be directly the html of the filter provided by the plugin.

The hook name should be singural, i suggest `dashboard_html_filter`

So in summary, the core should not care about testing the filter_id (new lines added from 1061 to 1062) as the current function ask for the html of a precise filter_id.

**I let other core developers give also their opinion, wait for it before doing any changes.**

**Comment by bessantoy** [comment]
I updated my solution so now the core doesn't take care about testing filter_id as the current function ask for the html of a precise filter_id. Waiting for your feedbacks.

Regards

**Comment by orthagh** [comment]
The last commit seems wrong :)

**Comment by bessantoy** [comment]
I reviewed my code following your advice, I implemented an abstract class representing a filter, all native filters extend this class and are in the \Dashboard\Filter folder, I modified the Filter.php file, the getAll() function now returns the classes representing the native filters as well as those retrieved by the DASHBOARD_FILTERS hook. Filter.php still contains the functions getForDashboard() and addForDashboard() which I have not modified
I also modified the Grid.php file to consider the new filter representation.

We can now add new filters in our plugin using the DASHBOARD_FILTERS hook like : 
 
`$PLUGIN_HOOKS[Hooks::DASHBOARD_FILTERS]['exemple'] = [
    MyFirstFilter::class,
    MySecondFilter::class
];`

and

class MyFirstFilter extends Dashboard\Filters\AbstractFilter
{
    public function getName(): string
    {
        return __("My first filter");
    }
    
    ...
}

Waiting for your feedbacks.

Regards,

Anthony from IT Gouvernance.

**Comment by bessantoy** [comment]
I cleaned my dirty commits and pushed one commit containing all my modifications

File: src/Dashboard/Filters/AbstractFilter.php
Comment (orthagh) [inline src/Dashboard/Filters/AbstractFilter.php:159]:
```suggestion
}

```

missing EOF

File: src/Plugin/Hooks.php
Comment (orthagh) [inline src/Plugin/Hooks.php:146]:
```suggestion  
```

File: src/Dashboard/Filters/DatesFilter.php
Comment (orthagh) [inline src/Dashboard/Filters/DatesFilter.php:38]: in the same namespace, an import is not required (you have the same thing in all below files, please fix also)

File: src/Dashboard/Filter.php
Comment (orthagh) [inline src/Dashboard/Filter.php:48]:
```suggestion
use Glpi\Dashboard\Filters\{
    DatesFilter,
    ItilCategoryFilter,
    LocationFilter,
    ManufacturerFilter,
    RequestTypeFilter,
    StateFilter,
    TicketTypeFilter,
    GroupTechFilter,
    UserTechFilter,
    DatesModFilter
};
```

Not a mandatory change, but fyi, you can have a shorter syntax for importing several classes of the same namespace)

**Review by orthagh — CHANGES_REQUESTED** [review]
Apart some sugar syntax remark and a miss, seems very good !

File: src/Plugin/Hooks.php
Comment (bessantoy) [reply src/Plugin/Hooks.php:146]: well, it's not

**Comment by bessantoy** [comment]
It seems clean now

**Comment by AdrienClairembault** [comment]
I think you also need to take these two functions into account : `getFiltersCriteria` and `getSearchFiltersCriteria`.
Without this two function I don't think any plugin filter can work properly.

The  `getFiltersCriteria` function explain to GLPI how to apply the filter (by adding the needed content to the SQL query).

image

On the other hand, `getSearchFiltersCriteria` is the function used to display the corrects search criteria that will be used when an user click on a widget and is taken to the search page.

image

There is also `getAllDasboardCards` but I'm not sure how these one are used, maybe @orthagh can help us here

image

**Comment by bessantoy** [comment]
So I considered the functions getSearchFiltersCriteria and getFiltersCriteria for the new representation of filters. Each Filter must now implement 2 new functions returning their crtieria and their searchcriteria

**Review by AdrienClairembault — CHANGES_REQUESTED** [review]
Looks good overall, I would just avoid using `$DB = DBConnection::getReadConnection();` in each filter as `DBConnection::getReadConnection()` can be a bit slow when using sql replication (see https://github.com/glpi-project/glpi/pull/12965).

Maybe initializing `$DB` once and then passing it as a parameter would be better for performances.

**Comment by bessantoy** [comment]
i fixed DatesFilter and DatesModFilter classes that i had badly implemented, i also fixed a syntax error.
I tested each filter one by one and everything works.

**Comment by orthagh** [comment]
CI says you still have some issues to fix :)

File: src/Dashboard/Filters/ItilCategoryFilter.php
Comment (cedric-anne) [inline src/Dashboard/Filters/ItilCategoryFilter.php:42]:
```suggestion
class ItilCategoryFilter extends AbstractFilter
```

**Review by orthagh — APPROVED** [review]

**Review by cconard96 — APPROVED** [review]

**Review by trasher — APPROVED** [review]
Seems good to me, I'll wait for @AdrienClairembault validation before merge since he follows this development and I did not

File: src/Dashboard/Filter.php
Comment (cedric-anne) [inline src/Dashboard/Filter.php:86]: Please verify here that corresponding plugin is active. Sometimes, plugins are registering their hook without being active.

File: src/Dashboard/Filter.php
Comment (cedric-anne) [inline src/Dashboard/Filter.php:85]: Previously, filters were returned as key => value. Now they will only be a set of values. It may break some plugins, like `advanceddashboards`.

Impacted piece of code:
```
'filters'    => array_keys(Filter::getAll()),
```

File: src/Dashboard/Filter.php
Comment (cedric-anne) [inline src/Dashboard/Filter.php:97]: Public methods should not be removed on a bugfixes version. At least, it should trigger a deprecation notice and maybe return a mocked value.

**Review by cedric-anne — CHANGES_REQUESTED** [review]

File: src/Dashboard/Filter.php
Comment (bessantoy) [reply src/Dashboard/Filter.php:85]: So do you mean that i should change this : `$filters = [
            DatesFilter::class,
            ...
        ];` to `$filters = [
            DatesFilter::getId() => DatesFilter::class,
           ...
        ];`

Comment (cedric-anne) [reply src/Dashboard/Filter.php:85]: I do not know how it is supposed to be used. Maybe @AdrienClairembault could help here.

File: src/Dashboard/Filter.php
Comment (cedric-anne) [inline src/Dashboard/Filter.php:155]:
```suggestion
        return "";
```

Comment (cedric-anne) [inline src/Dashboard/Filter.php:155]: Would this work?

```suggestion
        return DatesFilter::getHtml($values, $fieldname);
```

Comment (bessantoy) [reply src/Dashboard/Filter.php]: I think

File: src/Dashboard/Grid.php
Comment (AdrienClairembault) [inline src/Dashboard/Grid.php:1266]:
```suggestion
                    'Location',
```
I think this one should be reverted, does not seem related to filters.

File: src/Dashboard/Grid.php
Comment (AdrienClairembault) [inline src/Dashboard/Grid.php:1438]: Same for these 3, does not seem related to filters.

File: src/Dashboard/Provider.php
Comment (AdrienClairembault) [inline src/Dashboard/Provider.php:1641]: Search criteria are not working because they is one unneeded array level:

image

Can be fixed like this:

```suggestion
                array_push($s_criteria, ...$filter_criteria);
```

After fix:

image

**Review by AdrienClairembault — CHANGES_REQUESTED** [review]
There are also some inconsistencies with the the `getHtml` function.

In the abstract class, it does not have any parameters:

image

Most filters implements it with a single parameter:

image

The `Datefilter` implements it with two parameters :

image

The function seems to be called with only one parameter in the `Grid` class:

image

**Comment by AdrienClairembault** [comment]
One little functional issue that may be difficult to deal with.

Changing a custom filter does not trigger a refresh of the displayed cards.

This is because we only refresh cards that are marked as "impacted" by the filter:

image

These special `filter-*` classes are set depending on the data shown below, which may be difficult to set for each plugins filters as you need to do it for each existing cards

image

I think we should either always reload all cards or do some kind of generics checks (like it's done when adding the sql criteria where we check if a specific field exist in the table).

Any thoughts on this @orthagh @cedric-anne ?

**Review by AdrienClairembault — APPROVED** [review]
Ok for me in glpi main

**Comment by cedric-anne** [comment]
@bessantoy 

Could I push some commits to your branch, or would you prefer me to open a new PR based on a branch of mine, including your commits?
I want to push some minor changes on changelog, methods signatures, ... and it will probably faster for everyone if I push them directly.

**Comment by bessantoy** [comment]
> @bessantoy
> 
> Could I push some commits to your branch, or would you prefer me to open a new PR based on a branch of mine, including your commits? I want to push some minor changes on changelog, methods signatures, ... and it will probably faster for everyone if I push them directly.

Of course you can

**Comment by cedric-anne** [comment]
Replaced by #14717.

Thanks for the contribution.

**Pull request closed without merging**
